lib/services/access_list.go

```diff
@@ -218,47 +233,206 @@ func UnmarshalAccessListMember(data []byte, opts ...MarshalOption) (*accesslist.
 	// An opaque access denied error.
 	accessDenied := trace.AccessDenied("access denied")
+	// Is the Ownership inherited, or explicit?
+	inherited := false
 
 	// Is the supplied identity in the owners list?
-	ownerIdx := slices.IndexFunc(accessList.GetOwners(), func(owner accesslist.Owner) bool {
+	isOwner := slices.ContainsFunc(accessList.GetOwners(), func(owner accesslist.Owner) bool {
 		return owner.Name == identity.Username
 	})
-	if ownerIdx == -1 {
-		return accessDenied
+	if !isOwner {
+		err := recursiveIsAccessListOwnerCheck(ctx, members, identity, accessList)
+		if err != nil {
+			return false, accessDenied
+		}
+		inherited = true
 	}
 
 	// Does the supplied Identity meet the ownership requirements?
 	if !UserMeetsRequirements(identity, accessList.Spec.OwnershipRequires) {
-		return accessDenied
+		return inherited, accessDenied
 	}
 
 	// We've gotten through all the checks, so the user is an owner.
-	return nil
+	return inherited, nil
 }
 
 // AccessListMembershipChecker will check if users are members of an access list and
 // makes sure the user is not locked and meets membership requirements.
 type AccessListMembershipChecker struct {
-	members AccessListMemberGetter
+	members AccessListsAndMembersGetter
 	locks   LockGetter
 	clock   clockwork.Clock
 }
 
 // NewAccessListMembershipChecker will create a new access list membership checker.
-func NewAccessListMembershipChecker(clock clockwork.Clock, members AccessListMemberGetter, locks LockGetter) *AccessListMembershipChecker {
+func NewAccessListMembershipChecker(clock clockwork.Clock, members AccessListsAndMembersGetter, locks LockGetter) *AccessListMembershipChecker {
 	return &AccessListMembershipChecker{
 		members: members,
 		locks:   locks,
 		clock:   clock,
 	}
 }
 
-// IsAccessListMember will return true if the user is a member for the current list.
-func (a AccessListMembershipChecker) IsAccessListMember(ctx context.Context, identity tlsca.Identity, accessList *accesslist.AccessList) error {
+// getAccessListDynamicMembers returns a list of all dynamic members (nested lists) for a given Access List.
+func getAccessListDynamicMembers(ctx context.Context, clt AccessListsAndMembersGetter, entry string) ([]string, error) {
+	var pageToken string
+	var dynamicMembers []string
+
+	for {
+		var members []*accesslist.AccessListMember
+		var err error
+
+		members, pageToken, err = clt.ListAccessListMembers(ctx, entry, 0 /* default page size */, pageToken)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		for _, member := range members {
+			if member.Spec.MembershipKind == accesslist.MembershipKindList {
+				dynamicMembers = append(dynamicMembers, member.GetName())
+			}
+		}
+
+		if pageToken == "" {
+			break
+		}
+	}
+
+	return dynamicMembers, nil
+}
+
+func (a AccessListMembershipChecker) recursiveIsAccessListMemberCheck(ctx context.Context, identity tlsca.Identity, accessList *accesslist.AccessList) error {
```

Comment:
Godoc here would be helpful. Similar to the owners check function below, to explain that the passed in identity is checked for membership.